src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -411,78 +417,105 @@ export default memo(function ScatterPlotTab(props: ScatterPlotTabProps): ReactEl
-      const categories = dataset.getFeatureCategories(selectedFeatureName);
-      const isCategorical = categories !== null;
-      const colors = isCategorical
-        ? categoricalPalette.slice(0, categories.length)
-        : subsampleColorRamp(colorRamp, COLOR_RAMP_SUBSAMPLES);
-      const colorMinValue = isCategorical ? 0 : colorRampMin;
-      const colorMaxValue = isCategorical ? categories.length - 1 : colorRampMax;
-
-      // Make a bucket group for each ramp/palette color and for the out-of-range and outliers.
-      // index 0 = out of filter range, index 1 = outliers.
-      for (let i = 0; i < colors.length + 2; i++) {
-        let color;
-        if (i === 0) {
-          color = viewerConfig.outOfRangeDrawSettings.color;
-        } else if (i === 1) {
-          color = viewerConfig.outlierDrawSettings.color;
-        } else {
-          color = colors[i - 2];
-        }
-        traceDataBuckets.push({ x: [], y: [], objectIds: [], trackIds: [], color: color });
-      }
+      colors = subsampleColorRamp(colorRamp, COLOR_RAMP_SUBSAMPLES);
+    }
 
-      // Sort data into buckets
-      for (let i = 0; i < xData.length; i++) {
-        const isOutOfRange = false; // TODO: Implement out of range filtering
-        const isOutlier = dataset.outliers ? dataset.outliers[objectIds[i]] : false;
-        const objectId = objectIds[i];
-
-        let bucketIndex;
-        if (isOutOfRange) {
-          bucketIndex = 0;
-        } else if (isOutlier) {
-          bucketIndex = 1;
-        } else {
-          bucketIndex = getBucketIndex(featureData.data[objectId], colorMinValue, colorMaxValue, colors.length) + 2;
-        }
+    const colorMinValue = isCategorical ? 0 : colorRampMin;
+    const colorMaxValue = isCategorical ? categories.length - 1 : colorRampMax;
 
-        const bucket = traceDataBuckets[bucketIndex];
-        bucket.x.push(xData[i]);
-        bucket.y.push(yData[i]);
-        bucket.objectIds.push(objectIds[i]);
-        bucket.trackIds.push(trackIds[i]);
+    // Make a bucket group for each ramp/palette color and for the out-of-range and outliers.
+    // index 0 = out of filter range, index 1 = outliers.
+    const traceDataBuckets: TraceData[] = [];
+    for (let i = 0; i < colors.length + 2; i++) {
+      let color: TraceData["color"];
+      let marker: TraceData["marker"];
+      if (i === 0) {
+        color = `#${viewerConfig.outOfRangeDrawSettings.color.getHexString()}`;
+        marker = { ...markerConfig, ...markerConfig.outOfRange };
+      } else if (i === 1) {
+        color = `#${viewerConfig.outlierDrawSettings.color.getHexString()}`;
+        marker = { ...markerConfig, ...markerConfig.outliers };
+      } else {
+        color = `#${colors[i - 2].getHexString()}`;
+        marker = markerConfig;
       }
+      if (usingOverrideColor) {
+        color = `#${overrideColor.getHexString()}`;
+      }
+      traceDataBuckets.push({ x: [], y: [], objectIds: [], trackIds: [], color, marker });
+    }
```

Comment:
Might suggest unrolling "out of range" and "outlier" categories from this loop, rather than hitting an `if` for every iteration?

Reply:
Agree with the above.. also could preallocate the size of the traceDataBuckets array but I suppose this is not a performance bottleneck yet.  It's sufficiently readable for me the way it is too.